plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
-			p.logOnce["intel_uncore_frequency"] = err
-		}
-		return
+	p.needsTimeRelatedMsr = needsTimeRelatedMsr(p.CPUMetrics)
+
+	p.needsRapl = needsRapl(p.PackageMetrics)
+	p.needsMsrPackage = needsMsr(p.PackageMetrics)
+
+	p.cpuMetricsMap = map[string]func(telegraf.Accumulator, int, int, int){
+		cpuC0StateResidency.String():        p.addCPUC0StateResidency,
+		cpuC1StateResidency.String():        p.addCPUC1StateResidency,
+		cpuC3StateResidency.String():        p.addCPUC3StateResidency,
+		cpuC6StateResidency.String():        p.addCPUC6StateResidency,
+		cpuC7StateResidency.String():        p.addCPUC7StateResidency,
+		cpuBusyFrequency.String():           p.addCPUBusyFrequency,
+		cpuBusyCycles.String():              p.addCPUBusyCycles,
+		cpuC0SubstateC01Percent.String():    p.addCPUC0SubstateC01Percent,
+		cpuC0SubstateC02Percent.String():    p.addCPUC0SubstateC02Percent,
+		cpuC0SubstateC0WaitPercent.String(): p.addCPUC0SubstateC0WaitPercent,
+	}
+
+	p.packageMetricMap = map[string]func(telegraf.Accumulator, int){
+		packageCurrentPowerConsumption.String():     p.addCurrentPackagePower,
+		packageCurrentDramPowerConsumption.String(): p.addCurrentDramPower,
+		packageThermalDesignPower.String():          p.addThermalDesignPower,
+		packageCPUBaseFrequency.String():            p.addCPUBaseFrequency,
+		packageTurboLimit.String():                  p.addMaxTurboFreqLimits,
 	}
-	p.logOnce["intel_uncore_frequency"] = nil
-	p.readUncoreFreq("initial", socketID, die, acc)
-	p.readUncoreFreq("current", socketID, die, acc)
+
+	return nil
 }
 
-func (p *PowerStat) readUncoreFreq(typeFreq string, socketID string, die string, acc telegraf.Accumulator) {
-	fields := map[string]interface{}{}
-	if typeFreq == "current" {
-		if p.areCoreMetricsEnabled() && p.msr.isMsrLoaded() {
-			p.logOnce[socketID+"msr"] = nil
-			cpuID, err := p.GetCPUIDFromSocketID(socketID)
-			if err != nil {
-				p.Log.Debugf("Error while reading socket ID: %v", err)
-				return
-			}
-			actualUncoreFreq, err := p.msr.readSingleMsr(cpuID, msrUncorePerfStatusString)
-			if err != nil {
-				p.Log.Debugf("Error while reading %s: %v", msrUncorePerfStatusString, err)
-				return
-			}
-			actualUncoreFreq = (actualUncoreFreq & 0x3F) * 100
-			fields["uncore_frequency_mhz_cur"] = actualUncoreFreq
-		} else {
-			err := errors.New("Error while trying to read MSR (probably msr module was not loaded), uncore_frequency_mhz_cur metric will not be collected")
-			if val := p.logOnce[socketID+"msr"]; val == nil || val.Error() != err.Error() {
-				p.Log.Errorf("%v", err)
-				// Remember that specific error occurs for socketID to omit logging next time
-				p.logOnce[socketID+"msr"] = err
-			}
+// parsePackageMetrics takes a slice of package metrics. It ensures all metrics in the slice
+// are supported package-specific, and there are no duplicates. If metrics slice is nil, then
+// default package metrics are provided.
+func parsePackageMetrics(metrics []string) ([]string, error) {
```

Comment:
How about responding with the enum instead of the string?

Reply:
Since the type of the `PowerStat` field parsed from toml file is a slice of strings I think we could keep it as is, so there would be no need to convert from string to enum type.

Reply:
Added new commit [`98e961c4`](https://github.com/influxdata/telegraf/pull/14363/commits/98e961c4affec0767a1e95f9d3722d082952cc1c) which introduces changes to actually use the enum. It also introduces changes to helper functions to  use `switch` statement instead of defining a slice and loop through it. It addresses part of this other [`comment`](https://github.com/influxdata/telegraf/pull/14363#discussion_r1409197941).